libs/meta/tests/unit/callable_traits_tests.cpp

```diff
@@ -0,0 +1,253 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/macros.hpp"
+#include "meta/callable_traits.hpp"
+#include "same.hpp"
+
+#include "gmock/gmock.h"
+
+#include <cstddef>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+using namespace ::testing;
+
+namespace fetch {
+namespace meta {
+
+namespace {
+
+using EmptyArgs = std::tuple<>;
+#define ARGS std::size_t const &, int, std::vector<std::string> &
+using Args         = std::tuple<ARGS>;
+using Void         = void;
+using NonVoid      = std::string;
+using NoOwningType = void;
+
+#define ASSERTIONS(callable, return_type, args_tuple_type, owning_type)      \
+  {                                                                          \
+    using ClassType = owning_type;                                           \
+    using Callable  = decltype(callable);                                    \
+                                                                             \
+    using Traits = CallableTraits<Callable>;                                 \
+    static_assert(Same<return_type, Traits::ReturnType>(), "");              \
```

Comment:
Very subtle nitpicking: AFAIK, the Standard only mandates that the error message is included in compiler output. Would do no harm to double asserted expressions inside parentheses.

Reply:
Nice